### PR TITLE
New module: nsupdate

### PR DESCRIPTION
**Pull request by nerzhul**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
nsupdate

##### ANSIBLE VERSION

```
ansible 2.2.1.0
```

##### SUMMARY
Add nsupdate module to manage DNS records on a DNS server

This uses dnspython library

It's greatly inspired by https://github.com/mskarbek/ansible-nsupdate with some rework, better feedbacks and documentation addition

```
changed: [127.0.0.1 -> 172.23.1.1] => (item=A) => {"changed": true, "failed": false, "item": "A", "rc": 0, "rc_str": "NOERROR", "record": {"record": "server01", "ttl": 3600, "type": "A", "value": null, "zon
e": "webinfra.local."}}
changed: [127.0.0.1 -> 172.23.1.1] => (item=TXT) => {"changed": true, "failed": false, "item": "TXT", "rc": 0, "rc_str": "NOERROR", "record": {"record": "server02", "ttl": 3600, "type": "TXT", "value": null
, "zone": "webinfra.local."}}

```
Please note this module requires recent dnspython version with at least Python 2.6

### CODE REVIEW

File: lib/ansible/modules/network/nsupdate.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+            self.zone = module.params['zone']
+
+        if module.params['key_name']:
+            try:
+                self.keyring = dns.tsigkeyring.from_text({
+                    module.params['key_name']: module.params['key_secret']
+                })
+            except TypeError:
+                module.fail_json(msg='Missing key_secret')
+            except binascii_error as e:
```
Comment (mattclay) [inline lib/ansible/modules/network/nsupdate.py]: CI failure due to python 2.4 syntax error:

```
2017-02-07 17:14:00 Compiling ./lib/ansible/modules/network/nsupdate.py ...
2017-02-07 17:14:00   File "./lib/ansible/modules/network/nsupdate.py", line 178
2017-02-07 17:14:00     except binascii_error as e:
2017-02-07 17:14:00                            ^
2017-02-07 17:14:00 SyntaxError: invalid syntax
```

Comment (nerzhul) [reply lib/ansible/modules/network/nsupdate.py]: @mattclay will fix that today

**Comment by nerzhul** [comment]
Shippable test is wrong :( 

```
ERROR: Type comparison using type() found on line 197. Use isinstance() instead
```

refers to:
```
self.module.fail_json(msg='TSIG update error (%s): %s' % (type(e).__name__, str(e)))
```

**Comment by nerzhul** [comment]
Fixed, thanks @alikins

**Comment by angystardust** [comment]
@nerzhul will it work with a MS DNS server?

**Comment by nerzhul** [comment]
it uses DNS protocol updates, it should work if you allow ansible IP in your MS server to do DDNS updates (if i remember M$ doesn't support TSIG as authentication)

**Comment by angystardust** [comment]
It seems so...https://technet.microsoft.com/en-us/library/cc961412.aspx

**Comment by nerzhul** [comment]
then perfect, it support both bind & MS, please shipit :D

File: lib/ansible/modules/network/nsupdate.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'version': '1.0'}
+
+DOCUMENTATION = '''
+---
+module: nsupdate
+
+short_description: Setup OMAPI hosts.
```
Comment (jhawkesworth) [inline lib/ansible/modules/network/nsupdate.py]: Short description - this should at least mention DNS.  Not sure how OMAPI is relevant here?

**Comment by jhawkesworth** [comment]
This looks like it could be very useful - it would be great if the documentation mentioned that it can modify  MS DNS and perhaps link to
https://technet.microsoft.com/en-us/library/cc961412.aspx

**Comment by nerzhul** [comment]
Nice shot i missed this one.I copied omapi host documentation when writing nsupdate

**Comment by nerzhul** [comment]
@jhawkesworth fixed ! i added the MS tip too

**Comment by ravibhure** [comment]
+1 shipit

**Comment by nerzhul** [comment]
bot_status

File: lib/ansible/modules/network/nsupdate.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+            return 0
+
+
+def main():
+    tsig_algs = ['HMAC-MD5.SIG-ALG.REG.INT', 'hmac-md5', 'hmac-sha1', 'hmac-sha224',
+                 'hmac-sha256', 'hamc-sha384', 'hmac-sha512']
+
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(required=False, default='present', choices=['present', 'absent'], type='str'),
```
Comment (gundalow) [inline lib/ansible/modules/network/nsupdate.py:318]: Required option according to the docs.

**Review by gundalow — COMMENTED** [review]
Just one minor change needed

File: lib/ansible/modules/network/nsupdate.py (hunk 3 above)
Comment (nerzhul) [reply lib/ansible/modules/network/nsupdate.py:318]: @gundalow i fixed the doc thanks

**Comment by dagwieers** [comment]
It would be nice to have diff support as well, but as said nice-to-have.

shipit

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Merged, thanks for the PR and all reviewers.

**Comment by nerzhul** [comment]
Thanks !

**Pull request merged**